compiler/src/lnntoamm/Types.ts

```diff
@@ -242,6 +246,162 @@ class Builtin extends Type {
+    const indices = matrix.map(() => 0);
+    // keep it as for instead of while for debugging reasons
+    for (let i = 0; ; i++) {
+      const weight = indices.reduce((w, c) => w + c);
+      const argTys = matrix.map((options, ii) => options[indices[ii]]);
+      const fnsForWeight = fnsByWeight.get(weight) || [];
+      fnsForWeight.push(
+        ...fns.reduce((fns, fn) => {
+          const retTy = fn.resultTyFor(argTys, scope);
+          if (retTy === null) {
+            return fns;
+          } else {
+            return [...fns, [fn, retTy] as [Fn, Type]];
+          }
+        }, new Array<[Fn, Type]>()),
+      );
+      fnsByWeight.set(weight, fnsForWeight);
+      if (
+        indices.every(
+          (idxInDim, dimIdx) => idxInDim === matrix[dimIdx].length - 1,
+        )
+      ) {
+        break;
+      }
+      // now change the indices. This mostly works like binary addition,
+      // except each digit `i` is in base `j` where `j = matrix[i].length`
+      // so if `matrix[1].length === 1` then the carry for `indices[1]`
+      // is just passed to `matrix[0]`
+      indices.reduceRight((carry, _matIdx, idx) => {
+        indices[idx] += carry;
+        if (indices[idx] === matrix[idx].length) {
+          indices[idx] = 0;
+          return 1;
+        } else {
+          return 0;
+        }
+      }, 1);
+    }
+    const weights = Array.from(fnsByWeight.keys()).sort();
+    // weights is ordered lowest->highest so it's just a matter of
+    // appending the tuple at each weight to a list
+    const ret = weights.reduce((fns, weight) => {
+      let weightFns = fnsByWeight.get(weight);
+      weightFns = weightFns.filter(
+        ([weightedFn, _retTy]) =>
+          fns.findIndex(([fn, _retTy]) => fn === weightedFn) === -1,
+      );
+      return [...fns, ...weightFns];
+    }, new Array<[Fn, Type]>());
+    if (ret.length > originalLength) {
+      console.log('~~~ ERROR');
+      console.log('original:', originalLength);
+      console.log('retLengt:', ret.length);
+      console.log('args:    ', args);
+      console.log('matrix:  ', matrix);
+      console.log('byweight:', fnsByWeight);
+      console.log('indices: ', indices);
+      throw new Error('somehow got more options when fn selecting');
+    }
+    return ret;
+  }
+
+  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
+    if (ty instanceof FunctionType) {
+      return (
+        this.params.length === ty.params.length &&
+        this.params.every((param, ii) =>
+          param.compatibleWithConstraint(ty.params[ii], scope),
+        ) &&
+        this.retTy.compatibleWithConstraint(ty.retTy, scope)
+      );
+    } else if (ty instanceof OneOf || ty instanceof Generated) {
+      return ty.compatibleWithConstraint(this, scope);
+    } else {
+      return false;
+    }
+  }
+
+  constrain(to: Type, scope: Scope): void {
+    console.log(to);
+    TODO('figure out what it means to constrain a function type');
```

Comment:
Probably similar to how generic types would be constrained? You could treat function types like `Fn<A, R>` with A being the arg types and R being the return type, and then either also implement variadic generic types or just hardwire something like `Arg0`, `Arg1<A>`, `Arg2<A, B>`, ... to some large-ish maximum arg length.

Those simple generics could themselves be generated in the code and never exposed to the user so it wouldn't be *too* crazy to go up to something like `Arg99` if that is possible.

Reply:
right, but I'm waiting until I tackle #534 before I figure this out, since there aren't any cases where `FunctionType.constrain` is called (unless there's some weird bug)